### DISCUSSION TITLE
How to change nvim-cmp config?

### DISCUSSION THREAD

**Discussion by dro-sh** [NvChad/NvChad#3033] [Q&A] [answered]
I want to customize nvim-cmp by adding source for codeium, [plugin](https://github.com/Exafunction/codeium.nvim?tab=readme-ov-file#nvim-cmp). How to do it?
```lua
cmp.setup({
    -- ...
    sources = {
        -- ...
        { name = "codeium" }
    }
})
```

**Comment by T3sT3ro** [comment]
Moreso, how to do it with `config.cmp` file? Configuring cmp is a pretty cumbersome process I am failing at for the last days with errors like `failed to run 'config' for nvim-cmp, attempt to index local 's'`... It would be nice if there was a recipe in the nvchad for config in another file.

**Comment by siduck** [comment]
please read the docs for overriding plugin opts 😞

https://nvchad.com/docs/config/plugins

**Reply by T3sT3ro** [reply]
I've read the docs through and through, I wouldn't post without reading them first. The problem is not without simple config, but with config that takes into account the nvchad's cmp config. There is just too little info and examples for extending default cmp configurtion. I did try with various approaches with `otps`, `config`, static table, function, trying to append to a table returned by `require('nvchad.configs.cmp')` but it's just too confusing. I would need some more handholding in the docs, I think.

For example: how do I set up a command line completions like on the README of the nvim-cmp plugin without overriding nvchad's config:
```lua
-- `:` cmdline setup.
cmp.setup.cmdline(':', {
  mapping = cmp.mapping.preset.cmdline(),
  sources = cmp.config.sources({
    { name = 'path' }
  }, {
    {
      name = 'cmdline',
      option = {
        ignore_cmds = { 'Man', '!' }
      }
    }
  })
})
```

How to add a source of completions without completely overriding the table? And various other questions...

**Reply by siduck** [reply]
```lua

  {
    "hrsh7th/nvim-cmp",
    dependencies = {

      {
        "supermaven-inc/supermaven-nvim",
        opts = {},
      },

      {
        "hrsh7th/cmp-cmdline",
        event = "CmdlineEnter",
        config = function()
          local cmp = require "cmp"

          cmp.setup.cmdline("/", {
            mapping = cmp.mapping.preset.cmdline(),
            sources = { { name = "buffer" } },
          })

          cmp.setup.cmdline(":", {
            mapping = cmp.mapping.preset.cmdline(),
            sources = cmp.config.sources({ { name = "path" } }, { { name = "cmdline" } }),
            matching = { disallow_symbol_nonprefix_matching = false },
          })
        end,
      },
    },

    opts = function(_, opts)
      opts.sources[1].trigger_characters = { "-" }
      table.insert(opts.sources, 1, { name = "supermaven" })
    end,
  }
  ```
  
   thats my setup

**Reply by dro-sh** [reply]
Thanks, next works fine. I read docs about customization, but dont know how exactly add source to `sources` table

```lua
  {
    "hrsh7th/nvim-cmp",
    opts = function(_, opts)
      table.insert(opts.sources, { name = "codeium" })
    end,
  },
```

**Reply by siduck** [reply, marked as answer]
@dro-sh  this will work too but u need to copy default opts.sources cuz its not a dict and wont get merged

```lua
{ 
     "blabla cmp",
       opts = {
       
                   sources = {
                      do here
                     }
    }
}

**Reply by dro-sh** [reply]
You right, next config also works. Thanks!
```lua
  {
    "hrsh7th/nvim-cmp",
    opts = {
      sources = {
        name = "codeium",
      },
    },
  },
```

**Reply by dro-sh** [reply]
@siduck, I use above config, but I have lost all my snippets and suggestions :(

**Reply by siduck** [reply]
@dro-sh  because u didnt copy content from our default opts.sources!! and sources isnt a dict-table, its a list-table (array) so lazy.nvim wont merge it!!!!!!!!!!!!!!!!!! 

so include default sources too

**Comment by swahpy** [comment]
hi all, apologize to jump in. I have a quick question here about `nvim-cmp`, too. How to configure auto-completion in search mode using `/` or `?`?. I just setup it using below snippets, but did not work.
```lua
  -- Use buffer source for `/` and `?` (if you enabled `native_menu`, this won't work anymore).
  cmp.setup.cmdline({ '/', '?' }, {
    mapping = cmp.mapping.preset.cmdline(),
    sources = {
      { name = 'buffer' }
    }
  })
```
Thank you for any hints.

**Reply by siduck** [reply]
@swahpy my cmp config

```lua
{
    "hrsh7th/nvim-cmp",
    dependencies = {
      {
        "supermaven-inc/supermaven-nvim",
        opts = {},
      },
      {
        -- snippet plugin
        "L3MON4D3/LuaSnip",
        config = function(_, opts)
          require("luasnip").config.set_config(opts)

          local luasnip = require "luasnip"

          luasnip.filetype_extend("javascriptreact", { "html" })
          luasnip.filetype_extend("typescriptreact", { "html" })
          luasnip.filetype_extend("svelte", { "html" })

          require "nvchad.configs.luasnip"
        end,
      },

      {
        "hrsh7th/cmp-cmdline",
        event = "CmdlineEnter",
        config = function()
          local cmp = require "cmp"

          cmp.setup.cmdline("/", {
            mapping = cmp.mapping.preset.cmdline(),
            sources = { { name = "buffer" } },
          })

          cmp.setup.cmdline(":", {
            mapping = cmp.mapping.preset.cmdline(),
            sources = cmp.config.sources({ { name = "path" } }, { { name = "cmdline" } }),
            matching = { disallow_symbol_nonprefix_matching = false },
          })
        end,
      },
    },

    opts = function(_, opts)
      table.insert(opts.sources, 1, { name = "supermaven" })
    end,
  },
  ```

**Reply by swahpy** [reply]
thank you for your help. I was just not aware that I need to add `hrsh7th/cmp-cmdline` as the dependency. Now it works just fine.
